### ISSUE TITLE
`ptr_as_ptr` does not trigger inside macros

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust-clippy#8120]
### Summary

Casting a pointer with `as` does not trigger for code that comes from macro expansion even if that macro is local to the crate.

### Lint Name

ptr_as_ptr

### Reproducer

I tried this code:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=62c639bef9acba6c027e0e027999e78e

```rust
#![warn(clippy::all)]
#![warn(clippy::pedantic)]

extern "C" fn foo(ch: *const i8) {
    let _ = ch;
    todo!();
}

const F: &[u8] = b"x\0";

macro_rules! doit {
    () => {
        foo(F.as_ptr() as *const i8)
    }
}

pub fn main() {
    foo(F.as_ptr() as *const i8);
    doit!{}
}
```

I expected to see this happen: warning on the function call and on the function call inside the macro.

Instead, this happened: only one warning was issued:

```
    Checking playground v0.0.1 (/playground)
warning: `as` casting between raw pointers without changing its mutability
  --> src/main.rs:18:9
   |
18 |     foo(F.as_ptr() as *const i8);
   |         ^^^^^^^^^^^^^^^^^^^^^^^ help: try `pointer::cast`, a safer alternative: `F.as_ptr().cast::<i8>()`
   |
note: the lint level is defined here
  --> src/main.rs:2:9
   |
2  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::ptr_as_ptr)]` implied by `#[warn(clippy::pedantic)]`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 1.06s
```

### Version

```text
Rust 1.57.0 stable on playground
```

**Comment by rsmantini** [comment]
Hey, I would like to have go on this one.
@rustbot claim
